# Data preperation - create whole dataset

Check with the answer in discussion board: https://canvas.uts.edu.au/courses/26202/discussion_topics/378285

In [1]:
import os
import pandas as pd

In [2]:
# cwd = os.getcwd()  # Get the current working directory (cwd)
# files = os.listdir(cwd)  # Get all the files in that directory
# print("Files in %r: %s" % (cwd, files))

In [3]:
customer= pd.read_csv('/Users/lilywijara/Desktop/final/customers.csv', sep='|')

In [4]:
customer.shape

(1000, 15)

In [5]:
customer.head()

ssn               cc_num     first       last gender  \
0  115-04-4507        4218196001337  Jonathan    Johnson      M   
1  715-55-5575  4351161559407816183    Elaine     Fuller      F   
2  167-48-5821        4192832764832   Michael    Cameron      M   
3  406-83-7518     4238849696532874   Brandon   Williams      M   
4  697-93-1877     4514627048281480      Lisa  Hernandez      F   

                       street        city state    zip      lat     long  \
0        863 Lawrence Valleys      Ambler    PA  19002  40.1809 -75.2156   
1  310 Kendra Common Apt. 164      Leland    NC  28451  34.2680 -78.0578   
2            05641 Robin Port     Cordova    SC  29039  33.4275 -80.8857   
3      26916 Carlson Mountain  Birmingham    AL  35242  33.3813 -86.7046   
4             809 Burns Creek       Fargo    GA  31631  30.7166 -82.5801   

   city_pop                                    job         dob      acct_num  
0     32412                  Accounting technician  1959-10-03  888022315787  
1     27112                     Professor Emeritus  1963-06-07  917558277935  
2      4215   International aid/development worker  1973-05-30  718172762479  
3    493806                    Seismic interpreter  1942-12-26  947268892251  
4       559  Medical laboratory scientific officer  1939-05-22  888335239225

The number of customers is 1000.

In [36]:
# read the path
file_path = '/Users/lilywijara/Desktop/final'

#list all the files from the path
file_list = os.listdir(file_path)

In [5]:
file_list

['transactions_12.csv',
 'transactions_126.csv',
 'transactions_127.csv',
 'transactions_13.csv',
 'customers.csv',
 'transactions_39.csv',
 'transactions_11.csv',
 'transactions_119.csv',
 'transactions_131.csv',
 'transactions_125.csv',
 'transactions_124.csv',
 'transactions_130.csv',
 'transactions_118.csv',
 'transactions_10.csv',
 'transactions_38.csv',
 'transactions_14.csv',
 'transactions_28.csv',
 'transactions_120.csv',
 'transactions_108.csv',
 '.DS_Store',
 'transactions_109.csv',
 'transactions_121.csv',
 'transactions_29.csv',
 'transactions_15.csv',
 'transactions_17.csv',
 'transactions_123.csv',
 'transactions_122.csv',
 'transactions_16.csv',
 'transactions_59.csv',
 'transactions_65.csv',
 'transactions_71.csv',
 'transactions_7.csv',
 'transactions_6.csv',
 'transactions_70.csv',
 'transactions_64.csv',
 'transactions_58.csv',
 'transactions_72.csv',
 'transactions_4.csv',
 'transactions_66.csv',
 'transactions_99.csv',
 'transactions_98.csv',
 'transactions_67.csv

## Merge 2 file

In [16]:
transactions_1 = pd.read_csv('/Users/lilywijara/Desktop/final/transactions_1.csv', sep='|')
transactions_2 = pd.read_csv('/Users/lilywijara/Desktop/final/transactions_2.csv', sep='|')

In [18]:
transactions_master = transactions_1.merge(transactions_2, 
                   how = 'outer')
transactions_master

cc_num      acct_num                         trans_num  \
0      4065133387262473  950105810049  8bc87b4a938a94de66c73312a70b955e   
1      4065133387262473  950105810049  a091d0441f62b8195a005b698a9b25d7   
2      4065133387262473  950105810049  e458e6934892d3871fc49483f6744eab   
3      4065133387262473  950105810049  ec9286554b98b0f6272e7ad3e79ac46a   
4      4065133387262473  950105810049  ff9109f3fcfc870213143c7cd55dee1c   
...                 ...           ...                               ...   
69212  4646726368154185   44386431879  66fe860ddf8077e7bcc7d60606893c92   
69213  4646726368154185   44386431879  319088b8fc37909de223cd7beb6177c2   
69214  4646726368154185   44386431879  c70810355ef6878b86c34dbb6f8efbae   
69215  4646726368154185   44386431879  7cfcf2b4d3453c594612e825a823ad72   
69216  4646726368154185   44386431879  c2b1d6abaf4e734e7c6484311d936d82   

        unix_time        category    amt  is_fraud  \
0      1648587377     grocery_pos  43.25         0   
1      1651692302        misc_pos  98.94         0   
2      1642543681   gas_transport  53.33         0   
3      1650742723        misc_net  85.15         0   
4      1650746122        misc_net  61.35         0   
...           ...             ...    ...       ...   
69212  1612006901    shopping_pos   4.30         0   
69213  1632054304     food_dining   9.45         0   
69214  1639017649  health_fitness   6.53         0   
69215  1649415714            home   9.40         0   
69216  1637113534  health_fitness   4.41         0   

                           merchant  merch_lat  merch_long  
0      Olson, Montgomery and Vaughn  42.070264  -83.150048  
1                         Mills Ltd  43.586871  -83.310010  
2                   Reynolds-Gamble  42.564447  -83.768322  
3                      Rivera-Evans  43.403984  -82.304605  
4                    Santiago Group  41.733693  -82.953968  
...                             ...        ...         ...  
69212                Green and Sons  35.486360  -84.845374  
69213                Roberts-Guerra  34.293966  -85.474455  
69214                 Sanchez Group  34.625914  -86.131112  
69215                      Bell Ltd  34.566233  -85.004455  
69216                  Koch-Escobar  34.995944  -84.248997  

[69217 rows x 10 columns]

## For loop to merge 132 files

In [31]:
import pandas as pd

# create an empty list to store the transaction dataframes
transactions_list = []

# loop through the file names.
for i in range(0, 131):
    file_name = f'/Users/lilywijara/Desktop/final/transactions_{i}.csv'
    try:
        # read in the transaction data for each file and append it to the list
        transactions_list.append(pd.read_csv(file_name, sep='|'))
    except FileNotFoundError:
        # if the file is not found, print an error message and continue to the next file
        print(f"File {file_name} not found, skipping...")

# merge all the transaction dataframes into a single master transactions dataset
transactions_master = pd.concat(transactions_list, axis=0, ignore_index=True)

# print the merged dataset
transactions_master

cc_num      acct_num                         trans_num  \
0           4218196001337  888022315787  91ab12e73ef38206e1121e9648d2408d   
1           4218196001337  888022315787  071553d533a6822a4431c354c434ddcb   
2           4218196001337  888022315787  0cfad38ef15e4749eff68dc83f62c151   
3           4218196001337  888022315787  5782693d7c70f062f258cb30bfa8900f   
4           4218196001337  888022315787  35fd7db657d7e30dd608c37f7798186e   
...                   ...           ...                               ...   
4260899  2704390552958068  756767582811  1f091e64bd3f2382b3a4f90889329c32   
4260900  2704390552958068  756767582811  971aa63530f8f8fb1de28472e6aa53bc   
4260901  2704390552958068  756767582811  555941e364d1f591a6ae4088557f8ed3   
4260902  2704390552958068  756767582811  b756330bb5da439e28412f40e7a5f9e3   
4260903  2704390552958068  756767582811  d8b2caf59c06c36fea6980964f0a1fd6   

          unix_time        category    amt is_fraud  \
0        1558719550   gas_transport  69.12        0   
1        1569425519     grocery_pos  68.11        0   
2        1577205601        misc_net  40.35        0   
3        1571428238     grocery_pos  96.22        0   
4        1549840400   gas_transport  71.89        0   
...             ...             ...    ...      ...   
4260899  1662431076  health_fitness   8.21        0   
4260900  1665826484            home   4.79        0   
4260901  1662292853  health_fitness   1.99        0   
4260902  1670561624   personal_care   6.64        0   
4260903  1671856381          travel  12.74        0   

                            merchant  merch_lat  merch_long  
0                     Phillips Group  39.491416  -75.588522  
1                         Tucker Ltd  40.890319  -75.573359  
2                          Dixon PLC  39.244958  -74.475327  
3                     Lambert-Cooper  39.656925  -75.802342  
4                       Griffith LLC  40.313342  -74.220434  
...                              ...        ...         ...  
4260899                Anderson-King  33.831954 -116.587500  
4260900     Vang, Higgins and Decker  33.591525 -117.962834  
4260901                   Ibarra LLC  34.370334 -117.628700  
4260902  Casey, Gonzales and Mueller  33.690374 -117.301976  
4260903                 Cruz-Johnson  34.012253 -116.333944  

[4260904 rows x 10 columns]

The total number of transactions are 4,260,904.

## Merge the customer file with transactions

In [23]:
# merge the two datasets on the 'cc_num' and 'acct_num' columns
df = pd.merge(transactions_master, customer, on=['cc_num', 'acct_num'], how='left')

# print the merged dataset
print(df)

                   cc_num      acct_num                         trans_num  \
0           4218196001337  888022315787  91ab12e73ef38206e1121e9648d2408d   
1           4218196001337  888022315787  071553d533a6822a4431c354c434ddcb   
2           4218196001337  888022315787  0cfad38ef15e4749eff68dc83f62c151   
3           4218196001337  888022315787  5782693d7c70f062f258cb30bfa8900f   
4           4218196001337  888022315787  35fd7db657d7e30dd608c37f7798186e   
...                   ...           ...                               ...   
4260899  2704390552958068  756767582811  1f091e64bd3f2382b3a4f90889329c32   
4260900  2704390552958068  756767582811  971aa63530f8f8fb1de28472e6aa53bc   
4260901  2704390552958068  756767582811  555941e364d1f591a6ae4088557f8ed3   
4260902  2704390552958068  756767582811  b756330bb5da439e28412f40e7a5f9e3   
4260903  2704390552958068  756767582811  d8b2caf59c06c36fea6980964f0a1fd6   

          unix_time        category    amt is_fraud  \
0        1558719550 

## Check the final dataset

In [25]:
df.shape

(4260904, 23)

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4260904 entries, 0 to 4260903
Data columns (total 23 columns):
 #   Column      Dtype  
---  ------      -----  
 0   cc_num      object 
 1   acct_num    object 
 2   trans_num   object 
 3   unix_time   object 
 4   category    object 
 5   amt         float64
 6   is_fraud    object 
 7   merchant    object 
 8   merch_lat   float64
 9   merch_long  float64
 10  ssn         object 
 11  first       object 
 12  last        object 
 13  gender      object 
 14  street      object 
 15  city        object 
 16  state       object 
 17  zip         int64  
 18  lat         float64
 19  long        float64
 20  city_pop    int64  
 21  job         object 
 22  dob         object 
dtypes: float64(5), int64(2), object(16)
memory usage: 780.2+ MB


In [42]:
df.head()

cc_num      acct_num                         trans_num   unix_time  \
0  4218196001337  888022315787  91ab12e73ef38206e1121e9648d2408d  1558719550   
1  4218196001337  888022315787  071553d533a6822a4431c354c434ddcb  1569425519   
2  4218196001337  888022315787  0cfad38ef15e4749eff68dc83f62c151  1577205601   
3  4218196001337  888022315787  5782693d7c70f062f258cb30bfa8900f  1571428238   
4  4218196001337  888022315787  35fd7db657d7e30dd608c37f7798186e  1549840400   

        category    amt is_fraud        merchant  merch_lat  merch_long  ...  \
0  gas_transport  69.12        0  Phillips Group  39.491416  -75.588522  ...   
1    grocery_pos  68.11        0      Tucker Ltd  40.890319  -75.573359  ...   
2       misc_net  40.35        0       Dixon PLC  39.244958  -74.475327  ...   
3    grocery_pos  96.22        0  Lambert-Cooper  39.656925  -75.802342  ...   
4  gas_transport  71.89        0    Griffith LLC  40.313342  -74.220434  ...   

  gender                street    city state    zip      lat     long  \
0      M  863 Lawrence Valleys  Ambler    PA  19002  40.1809 -75.2156   
1      M  863 Lawrence Valleys  Ambler    PA  19002  40.1809 -75.2156   
2      M  863 Lawrence Valleys  Ambler    PA  19002  40.1809 -75.2156   
3      M  863 Lawrence Valleys  Ambler    PA  19002  40.1809 -75.2156   
4      M  863 Lawrence Valleys  Ambler    PA  19002  40.1809 -75.2156   

   city_pop                    job         dob  
0     32412  Accounting technician  1959-10-03  
1     32412  Accounting technician  1959-10-03  
2     32412  Accounting technician  1959-10-03  
3     32412  Accounting technician  1959-10-03  
4     32412  Accounting technician  1959-10-03  

[5 rows x 23 columns]

## Save into df.csv

In [29]:
# df.to_csv('df.csv')

## Reference:

Joshua, S. (2022, April 25). How to combine multiple CSV files using Python for your analysis. Medium. https://medium.com/@stella96joshua/how-to-combine-multiple-csv-files-using-python-for-your-analysis-a88017c6ff9e

Opallage, M. (2020, September 8). Answer to ‘python pandas error tokenizing data’. Stack Overflow. https://stackoverflow.com/a/63797557

***
**Subject**: 36106 Machine Learning Algorithms and Applications - Autumn 2023

**University**: University of Technology Sydney
***